README.md
# convert2pdf

## Description

It is a server with a route that allows you to convert **Html** and **Libre Office** documents to **Pdf**.

Supported formats:
* *odt*
* *ott*
* *oth*
* *odm*
* *otm*
* *odg*
* *otg*
* *odp*
* *otp*
* *ods*
* *ots*
* *odc*
* *odf*
* *odi*
* *docx*
* *doc*
* *html*

## How it works

### How to run the server on localhost

You will only need to start a container:

```
docker run -itd -p "127.0.0.1:8000:8000" geotrekce/convert2pdf
```

### How to make a request to the server

This server has only one route. To convert a valid document (**Python** exemple):

```
>>> from requests import post
>>> files = {'file': ('name.ext', open(path, 'rb'), 'mimetype')}
>>> response = post('http://127.0.0.1:8000/', files=files)
```


CHANGES.md
# CHANGELOG

## 0.0.1

**Html** and **Libre Office** documents to pdf conversion.


tests/settings.py
from os.path import dirname, join, realpath

ROOT = dirname(dirname(realpath(__file__)))
DOC_PATH = join(ROOT, 'tests', 'doc')
ODT_DOCS_PATH = join(DOC_PATH, 'odt')
HTML_DOCS_PATH = join(DOC_PATH, 'html')

EMPTY_ODT_PATH = join(ODT_DOCS_PATH, 'empty.odt')
ONE_PAGE_ODT_PATH = join(ODT_DOCS_PATH, 'one_page.odt')
TEN_PAGES_ODT_PATH = join(ODT_DOCS_PATH, 'ten_pages.odt')

EMPTY_HTML_PATH = join(HTML_DOCS_PATH, 'empty.html')
ONE_PAGE_HTML_PATH = join(HTML_DOCS_PATH, 'one_page.html')
TEN_PAGES_HTML_PATH = join(HTML_DOCS_PATH, 'ten_pages.html')


tests/test_exceptions.py
from io import BytesIO
from unittest import TestCase

from requests import post


class TestExceptions(TestCase):

    def test_bad_route(self):
        files = {}
        response = post('http://127.0.0.1:8000/bad/route', files=files)
        self.assertEqual(response.status_code, 404)
        self.assertEqual(response.content, b'{"errors":["Bad url."]}\n')

    def test_bad_files_param(self):
        files = {}
        response = post('http://127.0.0.1:8000/', files=files)
        self.assertEqual(response.status_code, 400)
        self.assertEqual(response.content, b'{"errors":["No file."]}\n')

    def test_bad_mime_type(self):
        files = {'file': ('doc.txt', BytesIO(b''), 'text/plain')}
        response = post('http://127.0.0.1:8000/', files=files)
        self.assertEqual(response.status_code, 400)
        self.assertEqual(
            response.content,
            b'{"errors":["Invalid mime type: text/plain is not supported."]}\n')


tests/test_odt2pdf_works.py
from unittest import TestCase

from requests import post
from tests.settings import (EMPTY_ODT_PATH, ONE_PAGE_ODT_PATH,
                            TEN_PAGES_ODT_PATH)


class TestOdt2PdfWorks(TestCase):

    def test_empty_odt_to_pdf(self):
        files = {'file': ('document.odt', open(EMPTY_ODT_PATH, 'rb'),
                          'application/vnd.oasis.opendocument.text')}
        response = post('http://127.0.0.1:8000/', files=files)
        self.assertEqual(response.status_code, 200)
        self.assertEqual(len(response.content), 1143)

    def test_one_page_odt_to_pdf(self):
        files = {'file': ('document.odt', open(ONE_PAGE_ODT_PATH, 'rb'),
                          'application/vnd.oasis.opendocument.text')}
        response = post('http://127.0.0.1:8000/', files=files)
        self.assertEqual(response.status_code, 200)
        self.assertEqual(len(response.content), 20890)

    def test_ten_pages_odt_to_pdf(self):
        files = {'file': ('document.odt', open(TEN_PAGES_ODT_PATH, 'rb'),
                          'application/vnd.oasis.opendocument.text')}
        response = post('http://127.0.0.1:8000/', files=files)
        self.assertEqual(response.status_code, 200)
        self.assertEqual(len(response.content), 94332)


convert2pdf/converter/converter.py
import os
import sys
from logging import INFO, Formatter, StreamHandler, getLogger
from tempfile import NamedTemporaryFile, gettempdir
from time import sleep

import uno
import unohelper
from com.sun.star.connection import NoConnectException
from com.sun.star.document.UpdateDocMode import QUIET_UPDATE
from convert2pdf.converter.utils import uno_props

logger = getLogger(__name__)
logger.setLevel(INFO)

formatter = Formatter('[%(asctime)s][%(levelname)s] %(message)s')

stream_handler = StreamHandler()
stream_handler.setLevel(INFO)
stream_handler.setFormatter(formatter)
logger.addHandler(stream_handler)


class Converter:
    """
    This convertor takes as optionnal parameters a host and port of a soffice server. It can be
    used to convert any LibreOffice document to a pdf.

    If you are working with the Linux operating system, you must install libreoffice and give the
    virtual environment access to the global site packages.

    If you are working with the MacOS operating system, you must install libreoffice and use the
    python it installs.
    """
    def __init__(self, server='127.0.0.1', port='2002'):
        connection = ("socket,host=%s,port=%s,tcpNoDelay=1;urp;StarOffice.ComponentContext" %
                      (server, port))

        # Do the LibreOffice component dance
        self.context = uno.getComponentContext()
        self.svcmgr = self.context.ServiceManager
        resolver = self.svcmgr.createInstanceWithContext("com.sun.star.bridge.UnoUrlResolver",
                                                         self.context)

        # Test for an existing connection
        unocontext = None
        while not unocontext:
            try:
                unocontext = resolver.resolve("uno:%s" % connection)
            except NoConnectException:
                logger.error('No soffice instance is running.')
                sleep(1)

        if not unocontext:
            sys.exit(1)

        # And some more LibreOffice magic
        unosvcmgr = unocontext.ServiceManager
        self.desktop = unosvcmgr.createInstanceWithContext("com.sun.star.frame.Desktop", unocontext)
        self.cwd = unohelper.systemPathToFileUrl(os.getcwd())

    def convert(self, inputfn):
        """
        This method takes a LibreOffice document as a byte object as parameter and returns the
        associated pdf file as a byte object.
        """
        document = None

        # Load inputfile
        inputprops = uno_props(Hidden=True, ReadOnly=True, UpdateDocMode=QUIET_UPDATE)

        if isinstance(inputfn, bytes):
            inputStream = self.svcmgr.createInstanceWithContext(
                "com.sun.star.io.SequenceInputStream", self.context)
            inputStream.initialize((uno.ByteSequence(inputfn),))
            inputprops += uno_props(InputStream=inputStream)
            inputurl = 'private:stream'
        else:
            inputfn_type = type(inputfn)
            logger.error('Incorrect type for the input parameter: %s unrecognized' % inputfn_type)
            sys.exit(1)

        document = self.desktop.loadComponentFromURL(inputurl, "_blank", 0, inputprops)
        if not document:
            logger.error('Could not open the document.')
            sys.exit(1)

        outputprops = uno_props(FilterName='writer_pdf_Export', Overwrite=True)

        outputf = NamedTemporaryFile()
        outputurl = unohelper.absolutize(gettempdir(), unohelper.systemPathToFileUrl(outputf.name))

        document.storeToURL(outputurl, outputprops)
        document.dispose()
        document.close(True)

        with open(outputf.name, 'rb') as read_file:
            pdf_content = read_file.read()
        if not pdf_content:
            logger.error('soffice error please check.')
            sys.exit(1)
        return pdf_content


convert2pdf/wsgi.py
from flask import Flask, jsonify, request
from weasyprint import HTML

from convert2pdf.converter import Converter
from convert2pdf.utils import LibreOfficeMimeTypes, send_file_from_string


class FlaskApp(Flask):

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.converter = Converter()


application = FlaskApp('convert2pdf')


@application.errorhandler(404)
def page_not_found(e):
    return jsonify(errors=['Bad url.']), 404


@application.route('/', methods=['POST'])
def to_pdf():
    if 'file' in request.files:
        f = request.files['file']
        mime_type = f.mimetype
        content = f.stream.read()

        if LibreOfficeMimeTypes.is_member(mime_type):
            pdf = application.converter.convert(content)
            return send_file_from_string(pdf, mime_type)
        elif mime_type == 'text/html':
            pdf = HTML(string=content.decode()).write_pdf()
            return send_file_from_string(pdf, mime_type)
        else:
            return (
                jsonify({'errors': ['Invalid mime type: %s is not supported.'
                                    % (mime_type, )]}),
                400
            )
    else:
        return (
            jsonify({'errors': ['No file.']}),
            400
        )
